File: 2021/day2/day2.hs
module Main where
import System.IO  
import Control.Monad
import Data.Foldable

readLines :: String -> IO [String]
readLines file = lines <$> readFile file

classify :: String -> (Char, Int)
classify line = (head wl, read wr)
   where (wl:wr:_) = words line


part1 :: [String] -> Int
part1 ls = h*d
   where
      (h,d) = foldl' go (0,0) (classify <$> ls)
      go (h,d) ('f', x) = (h+x, d)
      go (h,d) ('u', x) = (h, d-x)
      go (h,d) ('d', x) = (h, d+x)

part2 :: [String] -> Int
part2 ls = h*d
   where
      (h,d,_) = foldl' go (0,0,0) (classify <$> ls)
      go (h, d, a) ('f', x) = (h + x, d + a*x, a)
      go (h, d, a) ('d', x) = (h, d, a + x)
      go (h, d, a) ('u', x) = (h, d, a - x)


-- horiz = depth = aim = 0
-- lines.each do |line|
--   case line
--   when /forward (\d+)/ then horiz += $1.to_i ; depth += aim * $1.to_i
--   when /down (\d+)/    then aim += $1.to_i
--   when /up (\d+)/      then aim -= $1.to_i
--   end
-- end
-- puts "part2: #{horiz * depth}"

-- part2 ns = part1 $ zipWith3 (\a b c -> a + b + c) ns (tail ns) (tail $ tail ns)

main :: IO ()
main = do
   contents <- readLines "day2.txt"
   putStrLn $ "part1: " ++ (show $ part1 contents)
   putStrLn $ "part2: " ++ (show $ part2 contents)

File: 2019/day02/day2.rb
$* << 'input.txt'
opcodes = File.read($*[0]).chomp.split(',').map(&:to_i)

def run(opcodes, **map)
  opcodes = opcodes.dup
  map.each { opcodes[_1] = _2 }
  i = 0
  while true
    case opcodes[i]
    when 1 then opcodes[opcodes[i+3]] = opcodes[opcodes[i+1]] + opcodes[opcodes[i+2]]
    when 2 then opcodes[opcodes[i+3]] = opcodes[opcodes[i+1]] * opcodes[opcodes[i+2]]
    when 99 then break
    end
    i += 4
  end
  opcodes.first
end

puts run(opcodes, 1 => 12, 2 => 2)

100.times do |x|
  100.times do |y|
    if run(opcodes, 1 => x, 2 => y) == 19690720
      puts 100*x + y
      exit
    end
  end
end

